histogram/src/sliding_window.rs

```diff
@@ -0,0 +1,470 @@
+    /// * `n` must be greater than `a + b`
+    /// * `interval` in nanoseconds must fit within a `u64`
+    /// * `interval` must be at least 1 microsecond
+    pub fn new(
+        a: u8,
+        b: u8,
+        n: u8,
+        interval: core::time::Duration,
+        slices: usize,
+    ) -> Result<Self, BuildError> {
+        let config = Config::new(a, b, n)?;
+
+        Ok(Self {
+            config,
+            interval,
+            slices,
+            started: None,
+        })
+    }
+
+    /// Specify the start time for the histogram as a `UnixInstant`.
+    pub fn start(mut self, start: UnixInstant) -> Self {
+        self.started = Some(start);
+        self
+    }
+
+    /// Consume the builder and produce a sliding window histogram that uses
+    /// atomic operations.
+    pub fn build(self) -> Result<Histogram, BuildError> {
+        let (a, b, n) = self.config.params();
+
+        let mut h = Histogram::new(a, b, n, self.interval, self.slices)?;
+
+        // if we have some start time, we move the three time fields in the
+        // histogram as necessary
+        if let Some(start) = self.started {
+            if start < h.started {
+                let delta = h.started - start;
+                h.started -= delta;
+                h.tick_origin -= delta;
+                h.tick_at.fetch_sub(delta, Ordering::Relaxed);
+            } else {
+                let delta = start - h.started;
+                h.started += delta;
+                h.tick_origin += delta;
+                h.tick_at.fetch_add(delta, Ordering::Relaxed);
+            }
+        }
+
+        Ok(h)
+    }
+}
+
+impl Histogram {
+    /// Create a new histogram that stores values across a sliding window and
+    /// allows concurrent modification.
+    ///
+    /// # Parameters:
+    /// * `a` sets bin width in the linear portion, the bin width is `2^a`
+    /// * `b` sets the number of divisions in the logarithmic portion to `2^b`.
+    /// * `n` sets the max value as `2^n`. Note: when `n` is 64, the max value
+    ///   is `u64::MAX`
+    /// * `interval` is the duration of each discrete time slice
+    /// * `slices` is the number of discrete time slices
+    ///
+    /// # Constraints:
+    /// * `n` must be less than or equal to 64
+    /// * `n` must be greater than `a + b`
+    /// * `interval` in nanoseconds must fit within a `u64`
+    /// * `interval` must be at least 1 millisecond
+    pub fn new(
+        a: u8,
+        b: u8,
+        n: u8,
+        interval: core::time::Duration,
+        slices: usize,
+    ) -> Result<Self, BuildError> {
+        let now = Instant::now();
+        let started = UnixInstant::now();
+
+        let config = Config::new(a, b, n)?;
+
+        let mut live = Vec::with_capacity(config.total_bins());
+        live.resize_with(config.total_bins(), || AtomicU64::new(0));
+
+        let interval: u128 = interval.as_nanos();
+
+        if interval >= Duration::SECOND.as_nanos() as u128 * 3600 {
```

Comment:
this constraint isn't included in the doc/comment above

Reply:
Fixed. I think this can be resolved.